src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/FileBasedProgramsProjectSystem.cs

```diff
@@ -68,31 +66,47 @@ public FileBasedProgramsProjectSystem(
     {
         // There are a few cases here:
         //   1.  The document is a primordial document (either not loaded yet or doesn't support design time build) - it will be in the misc files workspace.
         //   2.  The document is loaded as a canonical misc file - these are always in the misc files workspace.
         //   3.  The document is loaded as a file based program - then it will be in the main workspace where the project path matches the source file path.
 
         // NB: The FileBasedProgramsProjectSystem uses the document file path (the on-disk path) as the projectPath in 'IsProjectLoadedAsync'.
-        var isLoadedAsFileBasedProgram = document.FilePath is { } filePath && await IsProjectLoadedAsync(filePath, cancellationToken);
+        var isLoadedAsFileBasedProgram = textDocument.FilePath is { } filePath && await IsProjectLoadedAsync(filePath, cancellationToken);
 
         // If this document has a file-based program syntactic marker, but we aren't loading it in a file-based programs project,
         // we need the caller to remove and re-add this document, so that it gets put in a file-based programs project instead.
         // See the check in 'LspWorkspaceManager.GetLspDocumentInfoAsync', which removes a document based on 'IsMiscellaneousFilesDocumentAsync' result,
         // then calls 'GetLspDocumentInfoAsync' again for the same request.
-        if (!isLoadedAsFileBasedProgram && VirtualProjectXmlProvider.IsFileBasedProgram(await document.GetTextAsync(cancellationToken)))
+        if (!isLoadedAsFileBasedProgram && VirtualProjectXmlProvider.IsFileBasedProgram(await textDocument.GetTextAsync(cancellationToken)))
             return false;
 
-        if (document.Project.Solution.Workspace == _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.Workspace)
+        if (textDocument.Project.Solution.Workspace == _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.Workspace)
+        {
+            // Do a check to determine if the misc project needs to be re-created with a new HasAllInformation flag value.
+            if (textDocument is Document document
+                && await document.GetSyntaxTreeAsync(cancellationToken) is { } syntaxTree)
+            {
+                var newHasAllInformation = await VirtualProjectXmlProvider.HasTopLevelStatementsAsync(syntaxTree, cancellationToken);
+                if (newHasAllInformation != document.Project.State.HasAllInformation)
+                {
+                    // TODO: replace this method and the call site in LspWorkspaceManager,
+                    // with a mechanism for "updating workspace state if needed" based on changes to a document.
+                    // Perhaps this could be based on actually listening for changes to particular documents, rather than whenever an LSP request related to a document comes in.
```

Comment:
will have to think about this.  We don't want to handle it inside the lsp text change notification itself as that intentionally does as little work as possible.  It could be done as an async handling triggered by lsp text changes - but would have to handle LSP requests only being eventually consistent in terms of which project its in.

but definitely space to refactor this to make it more clear how transitions happen.